### PR TITLE
Args v2

### PR DESCRIPTION
**Pull request by bitdivine**
# Motivation
We wish to stop baking canister IDs and URLs into the nns-dapp at build time.

# Changes
- Add a `<meta>` tag to the head of every index.html with:
  - A data attribute providing the own canister ID

# Tests
- TODO: How would this be tested using the shiny new e2e framework?

Tested manually:

* Download state: `curl -fL --retry 5 https://github.com/dfinity/snsdemo/releases/download/tip/state.zip > state.zip`
* Start dfx with the state:
```
$ dfx-state-install --state "$(dfx cache show)/state.zip"
replace .config/dfx/networks.json? [y]es, [n]o, [A]ll, [N]one, [r]ename: n
replace .config/dfx/identity/snsdemo8/identity.pem? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
replace .config/dfx/identity/snsdemo8/neurons/local? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
replace .config/dfx/identity/snsdemo8/identity.json? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
replace .config/dfx/identity/ident-1/identity.pem? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
replace .config/dfx/identity/ident-1/identity.json? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
Running dfx start for version 0.13.1
[SNIP]
```
* Use docker builds for the nns-dapp: `jq '.canisters["nns-dapp"].build="./scripts/docker-build"' dfx.json | sponge dfx.json`
* Install the nns-dapp: `dfx deploy nns-dapp`
* Look at the source code of the nns-dapp in the browser:
```
...
<link rel="modulepreload" ...>
  <meta name="nns-dapp-vars" data-own-canister-id="qsgjb-riaaa-aaaaa-aaaga-cai"></head>
  <body>
    <div>
...
```

### CODE REVIEW

File: rs/backend/src/arguments.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    /// Values that are to be set in the web front end, by injecting them into Javascript.
+    args: Vec<(String, String)>,
+}
+
+thread_local! {
+  pub static CANISTER_ARGUMENTS: RefCell<CanisterArguments> = RefCell::new(CanisterArguments::default().with_own_canister_id());
+}
+
+impl CanisterArguments {
+    /// HTML to be appended onto _every_ index.html
```
Comment (bitdivine) [inline rs/backend/src/arguments.rs:16]: Note: The HTML body seems to be overwritten when svelte loads, but the head seems to be preserved.  Hence the use of a meta tag in the head to provide args.

File: rs/backend/Cargo.toml [hunk 2]
```diff
@@ -50,6 +50,7 @@ icp-ledger = { git = "https://github.com/dfinity/ic", rev = "21d3a113dc765b5e580
 on_wire = { git = "https://github.com/dfinity/ic", rev = "21d3a113dc765b5e5803d9ad4dc68e14880ae6a0" }
 registry-canister = { git = "https://github.com/dfinity/ic", rev = "21d3a113dc765b5e5803d9ad4dc68e14880ae6a0" }
 idl2json = { git = "https://github.com/dfinity/idl2json", rev="c983dea6fc5666912218e7babf2b6b2f8b5648a4" }
+flate2 = "1.0.25"
```
Comment (lmuntaner) [inline rs/backend/Cargo.toml:53]: Did you check `cargo audit` to see whether there are any security vulnerability with this dependency?

**Review by lmuntaner — COMMENTED** [review]
I think it's going in the right direction.

Can you deploy it to a testnet to check it out?

I guess that we'll need to change the `deploy.sh` script, right?

**Comment by bitdivine** [comment]
> I guess that we'll need to change the deploy.sh script, right?

Not for this PR, as the arguments are not set here.

**Review by lmuntaner — APPROVED** [review]
Thanks!

**Pull request merged**